### PR TITLE
limit and clear cache obligations opportunistically

### LINKED ISSUE #43613: rustc runs out of memory when compiling trait heavy code
Been working on an [update to combine](https://github.com/Marwes/combine/pull/105) and ran into https://github.com/rust-lang/rust/issues/43132. Since the fix has been merged I updated using rustup but now rustc runs out of memory when compiling instead.

This branch reproduces the issue https://github.com/Marwes/combine/tree/ice_allocator_exhausted. I haven't bothered to reduce to a minimal production given that it is likely related to the slow down mentioned in #43132 and removing code will just make it compile and make the problem less pronounced (possibly the minimal example here reproduces the same problem as well https://github.com/rust-lang/rust/pull/32062/files#diff-6a2bc45424fb5e8e0f7eb99f2e05f9cf). 

```
$ rustc --version --verbose
rustc 1.21.0-nightly (dd53dd5f9 2017-08-01)
binary: rustc
commit-hash: dd53dd5f9e21dce1fbc06b7f9f451d1009bdcfd8
commit-date: 2017-08-01
host: x86_64-pc-windows-msvc
release: 1.21.0-nightly
LLVM version: 4.0
```

### PR DESCRIPTION
**Pull request by nikomatsakis**
Keeping **all** the obligations for every projection is wasteful of
memory and compilation time. We only really care about those
subobligations that may inform the result of the projection (i.e., may
help to resolve any inference variables that appear within).
Therefore, we can clear the subobligations from the cache that don't
potentially affect the result of the projection. On every cache hit,
we also take the opportunity to check if the type variables have been
resolved *yet* and, if so, clear out the pending obligations.

Fixes #43613.

r? @arielb1 

NB -- not sure how to test for this. Probably we should add the #43613 test case to perf.

### CODE REVIEW

**Comment by arielb1** [comment]
Sure. We probably want to add the test case to perf-collector. 

I think my real worry here soundness-wise is that we could look up and "confirm" a `<T as Trait>::Ty = X` projection while `T: Trait` does not hold. I think the way we are supposed to handle this today is to make sure `T: Trait` is always checked "beforehand" (say, as a part of a WF-checking pass).

Maybe just add an additional `T: Trait` subobligation? That should be "deduplicated" more easily - we already have plenty of caches for *that*, and it will probably prevent some odd edge case.

We need to have a better model of soundness + caching.

**Comment by nikomatsakis** [comment]
@arielb1

I think the argument for why that cannot happen is that `project` doesn't, generally, succeed, unless  either:

- we know "directly" that `T: Trait` would be satisfied (or, at least, it generates sub-obligations that are sufficient to show that `T: Trait` would be satisfied, which must themselves be proven of course) -- e.g., clearly when it builds on `select`, that is the case. 
- there is some where-clause that tells us the projection result which must be validated elsewhere. 

This makes a sort of inductive argument. Are there cases you're thinking of that don't fall into these two categories?

In any case, I'm not quite sure what you're proposing regarding this "extra" `T: Trait` obligation -- are you saying that we should just uniformly add such an obligation for every cache hit? (In addition to keeping the obligations that this PR keeps, i.e., those that may inform the result?)

**Comment by nikomatsakis** [comment]
> We need to have a better model of soundness + caching.

I don't disagree; I still think we can move towards a more chalk-like model, where the "trait system impl" doesn't handle things like "traits and projections" directly, but rather just interprets goals that it doesn't understand as deeply. We can then uniformly insert caching here that would replace and subsume the existing patchwork of caching. It'd be nice to prototype this more directly in chalk -- right now the impl is intentionally kept "naive", but I think it'd be good to have a less naive, caching variant to show how it could work.

**Comment by nikomatsakis** [comment]
@arielb1 does that last commit look like what you had in mind? I tested it -- it still preserves the memory/performance benefits on the test case in question, at least:

```
    MB
22.70^                                             #
     |            :@::::::::::::::::@:::::::::::@::#::@::::::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     |            :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | ::::::@:@:::@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
     | :::: :@:@: :@::::: ::: ::::::@: : :: ::::@: #::@:: :::::::::::@:::::@::
   0 +----------------------------------------------------------------------->Gi
     0                                                                   67.21
```

File: src/librustc/traits/project.rs [hunk 1]
```diff
@@ -562,6 +581,71 @@ fn opt_normalize_projection_type<'a, 'b, 'gcx, 'tcx>(
...
+    NormalizedTy { value: result.value, obligations }
+}
+
+/// Whenever we give back a cache result for a projection like `<T as
+/// Trait>::Item ==> X`, we *always* include the obligation to prove
+/// that `T: Trait` (we may also include some other obligations). This
+/// may or may not be necessary -- in principle, all the obligations
+/// that must be proven to show that `T: Trait` were also returned
+/// when the cache was first populated.  But there is a vague concern
+/// that perhaps someone would not have proven those, but also not
```
Comment (arielb1) [inline src/librustc/traits/project.rs]: My concern wasn't about snapshot misuse, it was more about cache(/other "concurrent" operation) pollution.

**Comment by nikomatsakis** [comment]
@bors p=1

I'm marking this as high priority because it fixes a P-high performance regression and because I would like to backport it.

**Comment by nikomatsakis** [comment]
Marking as beta-accepted. Fairly targeted patch, and fixes a major regression. It *is* my PR though so it wouldn't hurt to have people check it out.

cc @rust-lang/compiler

**Comment by nikomatsakis** [comment]
Backporting now.

**Pull request merged**